Allow multiple input datasets in jupyter interactive tool

martenson opened the pull request:
Also works with a collection of type "list". 

+ fix a bug with saving notebooks on exit that had non-default name
+ remove the confusion with naming notebooks after inputs
+ bump version

tested on release_21.01

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

martenson committed:
allow multiple inputs in jupyter IT

+ fix a bug with saving notebooks on exit that had non-default name
+ remove the confusion with naming notebooks after inputs

martenson committed:
bump version

martenson committed:
respect the cheetah style

mvdbeek commented on tools/interactive/interactivetool_jupyter_notebook.xml:
```diff
@@ -17,12 +17,22 @@
+                #set $cleaned_name = $cleaned_name + '_' + ''.join(random.choice(string.ascii_lowercase) for _ in range(3))
+            #end if
+            #set $cleaned_names += [$cleaned_name]
+            ln -sf '$file' './jupyter/data/${cleaned_name}' &&
```
You might want to add the extension as well
```suggestion
            ln -sf '$file' './jupyter/data/${cleaned_name}.${file.ext}' && 
```

martenson replied:
I like that idea.

bgruening commented on tools/interactive/interactivetool_jupyter_notebook.xml:
```diff
@@ -17,12 +17,22 @@
-        #set $cleaned_name = re.sub('[^\w\-\.]', '_', str($input.element_identifier))
-        ln -sf '$input' './jupyter/data/${cleaned_name}' &&
+        #set $cleaned_names = []
+        #for $file in $input
```
you could use enumerate() here and then get a unique number?

martenson replied:
Do we want the order of inputs to be significant though?

martenson replied:
I changed the names to be prefixed with an index according to your suggestion.

mvdbeek commented on tools/interactive/interactivetool_jupyter_notebook.xml:
```diff
@@ -17,12 +17,22 @@
-        ln -sf '$input' './jupyter/data/${cleaned_name}' &&
+        #set $cleaned_names = []
+        #for $file in $input
+            #set $cleaned_name = re.sub('[^\w\-\.]', '_', str($file.element_identifier))
```
```suggestion
            #set $cleaned_name = re.sub('[^\w\-\.\s]', '_', str($file.element_identifier))
```
Since we're running the container we can be pretty sure whitespace is ok

martenson replied:
Cool, I've included this also in the notebook's name when running on existing

martenson commented on tools/interactive/interactivetool_jupyter_notebook.xml:
```diff
@@ -35,18 +45,17 @@
-            cp '$mode.ipynb' ./${cleaned_name}.ipynb &&
-            jupyter trust ./${cleaned_name}.ipynb &&
-
+            #set $noteboook_name = re.sub('[^\w\-\.]', '_', str($mode.ipynb.element_identifier))
```
```suggestion
            #set $noteboook_name = re.sub('[^\w\-\.\s]', '_', str($mode.ipynb.element_identifier))
```

martenson committed:
Apply suggestions from code review

include file ext with a name, allow spaces in input datasets and notebook names

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

martenson committed:
prefix all datasets with their index

martenson committed:
Merge branch 'jupyter-multiple' of github.com:martenson/galaxy into jupyter-multiple

martenson committed:
properly quote var expansion, since it can contain spaces now
